### ISSUE TITLE
Exception when importing hy.core.macros from Python

### ISSUE THREAD

**Issue by timmartin** [hylang/hy#1016]
If I open a Python 3.4 REPL and type `import hy.core.macros`, it throws an exception:

```
>>> import hy.core.macros
Traceback (most recent call last):
  File "/home/timm/dev/hy/hy/macros.py", line 184, in macroexpand_1
    obj = wrap_value(m(*ntree[1:], **opts))
  File "<eval_body>", line 1, in _hy_anon_fn_2
NameError: name 'is_empty' is not defined

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<frozen importlib._bootstrap>", line 2237, in _find_and_load
  File "<frozen importlib._bootstrap>", line 2226, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 1191, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 1161, in _load_backward_compatible
  File "/home/timm/dev/hy/hy/importer.py", line 188, in load_module
    self.path)
  File "/home/timm/dev/hy/hy/importer.py", line 76, in import_file_to_module
    _ast = import_file_to_ast(fpath, module_name)
  File "/home/timm/dev/hy/hy/importer.py", line 68, in import_file_to_ast
    return hy_compile(import_file_to_hst(fpath), module_name)
  File "/home/timm/dev/hy/hy/compiler.py", line 2596, in hy_compile
    result = compiler.compile(tree)
  File "/home/timm/dev/hy/hy/compiler.py", line 433, in compile
    ret = self.compile_atom(_type, tree)
  File "/home/timm/dev/hy/hy/compiler.py", line 425, in compile_atom
    ret = _compile_table[atom_type](self, atom)
  File "/home/timm/dev/hy/hy/compiler.py", line 627, in compile_raw_list
    ret = self._compile_branch(entries)
  File "/home/timm/dev/hy/hy/compiler.py", line 489, in _compile_branch
    return _branch(self.compile(expr) for expr in exprs)
  File "/home/timm/dev/hy/hy/compiler.py", line 316, in _branch
    results = list(results)
  File "/home/timm/dev/hy/hy/compiler.py", line 489, in <genexpr>
    return _branch(self.compile(expr) for expr in exprs)
  File "/home/timm/dev/hy/hy/compiler.py", line 433, in compile
    ret = self.compile_atom(_type, tree)
  File "/home/timm/dev/hy/hy/compiler.py", line 425, in compile_atom
    ret = _compile_table[atom_type](self, atom)
  File "/home/timm/dev/hy/hy/compiler.py", line 2006, in compile_expression
    ret = self.compile_atom(fn, expression)
  File "/home/timm/dev/hy/hy/compiler.py", line 425, in compile_atom
    ret = _compile_table[atom_type](self, atom)

  File "/home/timm/dev/hy/hy/compiler.py", line 366, in checker
    return fn(self, expression)
  File "/home/timm/dev/hy/hy/compiler.py", line 2450, in compile_macro
    ret = self._compile_time_hack(new_expression)
  File "/home/timm/dev/hy/hy/compiler.py", line 2428, in _compile_time_hack
    self.module_name)
  File "/home/timm/dev/hy/hy/importer.py", line 116, in hy_eval
    _ast, expr = hy_compile(hytree, module_name, get_expr=True)
  File "/home/timm/dev/hy/hy/compiler.py", line 2596, in hy_compile
    result = compiler.compile(tree)
  File "/home/timm/dev/hy/hy/compiler.py", line 433, in compile
    ret = self.compile_atom(_type, tree)
  File "/home/timm/dev/hy/hy/compiler.py", line 425, in compile_atom
    ret = _compile_table[atom_type](self, atom)
  File "/home/timm/dev/hy/hy/compiler.py", line 2006, in compile_expression
    ret = self.compile_atom(fn, expression)
  File "/home/timm/dev/hy/hy/compiler.py", line 425, in compile_atom
    ret = _compile_table[atom_type](self, atom)
  File "/home/timm/dev/hy/hy/compiler.py", line 366, in checker
    return fn(self, expression)
  File "/home/timm/dev/hy/hy/compiler.py", line 1418, in compile_decorate_expression
    fn = self.compile(expr.pop(-1))
  File "/home/timm/dev/hy/hy/compiler.py", line 433, in compile
    ret = self.compile_atom(_type, tree)
  File "/home/timm/dev/hy/hy/compiler.py", line 425, in compile_atom
    ret = _compile_table[atom_type](self, atom)
  File "/home/timm/dev/hy/hy/compiler.py", line 2006, in compile_expression
    ret = self.compile_atom(fn, expression)
  File "/home/timm/dev/hy/hy/compiler.py", line 425, in compile_atom
    ret = _compile_table[atom_type](self, atom)
  File "/home/timm/dev/hy/hy/compiler.py", line 366, in checker
    return fn(self, expression)
  File "/home/timm/dev/hy/hy/compiler.py", line 2301, in compile_function_def
    body = self._compile_branch(expression)
  File "/home/timm/dev/hy/hy/compiler.py", line 489, in _compile_branch
    return _branch(self.compile(expr) for expr in exprs)
  File "/home/timm/dev/hy/hy/compiler.py", line 316, in _branch
    results = list(results)
  File "/home/timm/dev/hy/hy/compiler.py", line 489, in <genexpr>
    return _branch(self.compile(expr) for expr in exprs)
  File "/home/timm/dev/hy/hy/compiler.py", line 433, in compile
    ret = self.compile_atom(_type, tree)
  File "/home/timm/dev/hy/hy/compiler.py", line 425, in compile_atom
    ret = _compile_table[atom_type](self, atom)
  File "/home/timm/dev/hy/hy/compiler.py", line 1992, in compile_expression
    expression = macroexpand(expression, self)
  File "/home/timm/dev/hy/hy/macros.py", line 149, in macroexpand
    tree = macroexpand_1(tree, compiler)
  File "/home/timm/dev/hy/hy/macros.py", line 191, in macroexpand_1
    raise HyMacroExpansionError(tree, msg)
hy.errors.HyMacroExpansionError:   File "/home/timm/dev/hy/hy/core/macros.hy", line 110, column 3

    (cond
    ^----
     [(odd? (len args))
  ---------------------
      (macro-error args "`for' requires an even number of args.")]
  ----------------------------------------------------------------
     [(empty? body)
  -----------------
      (macro-error None "`for' requires a body to evaluate")]
  -----------------------------------------------------------
     [(empty? args) `(do ~@body ~@belse)]
  ---------------------------------------
     [(= (len args) 2) `(for* [~@args] (do ~@body) ~@belse)]
  ----------------------------------------------------------
     [true
  --------
       (let [alist (cut args 0 nil 2)]
  -----------------------------------
        `(for* [(, ~@alist) (genexpr (, ~@alist) [~@args])] (do ~@body) ~@belse))]))
  --------------------------------------------------------------------------------^
HyMacroExpansionError: b'expanding `cond\': NameError("name \'is_empty\' is not defined",)'
```

This is with the current development branch (f9cf39ee738c). I don't see this with the current stable release (0.11.1).

**Comment by Kodiologist** [comment]
The bug still stands as of 11481ab, and it's pretty darn mysterious.

**Comment by kirbyfan64** [comment]
This is basically a bit of a circular dependency issue: `hy.core.macros` depends on helpers from `hy.core.language`, which in turn depends on `hy.core.macros`, etc.

**Comment by kirbyfan64** [comment]
Not exactly sure how to solve this cleanly...

**Comment by tuturto** [comment]
This might have changed since I last checked, but I thought there is no way to use Hy macros from python code? So even if importing would work, the result wouldn't be useful on Python. Importing gives you set of anonymous functions that when called return HyExpression.

**Comment by kirbyfan64** [comment]
Maybe importing `hy.core.macros` directly should just be an error...

**Comment by timmartin** [comment]
FWIW, I don't remember why I originally hit this, but it was related to my Hy plugin for pycov (https://github.com/timmartin/hy-coverage).

The original problem was a little more complicated than importing `hy.core.macros` directly, but I was importing one of the Hy modules that implicitly led to importing `hy.core.macros`, which failed. I simplified the problem down to `hy.core.macros` but couldn't get any further with it.

The coverage plugin needs to include every module that's included (directly or transitively) by the code that it's assessing coverage on. However, I don't seem to be hitting this any more, perhaps because of another fix I made to the way my code imports modules (I forget the details), so I assume I no longer have a need to import `hy.core.macros`.
